### PR TITLE
docs: add a central glossary

### PR DESCRIPTION
**Pull request by sennyeya**
## Hey, I just made a Pull Request!

#22001, builds on the work added by @dwelsch-esi in [his review of the Backstage docs](https://github.com/cncf/techdocs/blob/2abfd2e66358215dd0911c3fdfb7ed2364cf758f/assessments/0008-backstage/backstage-glossary.md) (thanks! 🙏 ).  

adds a central glossary and tries to move items from existing glossaries into the central glossary. Some `concepts.md` files are more than just a glossary and end up in a weird not 1:1 with the central glossary. Where possible, I've rerouted definitions in `concept` files to the glossary and tried to leave deeper explanations in the concepts files.

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: docs/references/glossary.md [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+
+Existing search technology that [Backstage Search](#search) can take advantage of through its modular design. Lunr is the default search in [Backstage Search](#search). Part of [Backstage Search](#search).
+
+## Software Catalog
+
+A Backstage plugin that provides a framework to keep track of ownership and metadata for any number and type of software [components](#component). A core feature of Backstage.
+
+## Software Templates
+
+A Backstage plugin with which to create [components](#component) in Backstage. A core feature of Backstage.
```
Comment (freben) [inline docs/references/glossary.md:282]: The tagline could be tweaked a bit down the line here :)

This is also a good opportunity to add the Scaffolder term and cross linking the two, since we confusingly refer to both in different contexts.

File: docs/references/glossary.md [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+## Conditional decisions
+
+[Rules](#permission-rules) need additional data before they can be used in a decision. Once a [rule](#permission-rule) is bound to relevant information it forms a [condition](#condition). Conditional decisions tell the [permission framework](#permission) to delegate evaluation to the [plugin](#plugin) that owns the corresponding [resource](#permission-resource). Permission requests that result in a conditional decision are allowed if all of the provided conditions evaluate to be true.
+
+## Contributor
+
+A volunteer who helps to improve an OSS product such as Backstage. This volunteer effort includes coding, testing, technical writing, user support, and other work. A [user role](#user-role).
+
+## Decorators
```
Comment (freben) [inline docs/references/glossary.md:90]: Bit of a scoping / context problem here when reading it in isolation - could be `Search Decorator` or `Decorator (Search plugin)` or something like that? To immediately upfront know what we're talking about

**Review by freben — COMMENTED** [review]
Alright some early comments to get started

File: docs/references/glossary.md (hunk 2 above)
Comment (sennyeya) [reply docs/references/glossary.md:90]: I tried Search decorators and didn't love it as it clumps all search terms next to each other, Decorators (search plugin) seems like a good alternative. The k8s glossary has an idea of tags that we could emulate as well, https://kubernetes.io/docs/reference/glossary/?all=true.

File: docs/references/glossary.md [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+
+Two important responsibilities of any authorization system are to decide if a user can do something, and to enforce that decision. In the Backstage permission framework, policies are responsible for decisions and plugins (typically backends) are responsible for enforcing them.
+
+## Popup
+
+A separate browser window opened on top of the previous one.
+
+## Procedure
+
+A set of actions that accomplish a goal, usually as part of a [use case](#Use-Case). A procedure can be high-level, containing other procedures, or can be as simple as a single [task](#Task).
```
[the remark this answers is not in the document]
Comment (sennyeya) [reply docs/references/glossary.md:248]: updated in https://github.com/backstage/backstage/pull/22542/commits/e221a0b285500e5c1ce269aefdf7376f7a3ddb63

File: docs/references/glossary.md [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+## Component (catalog plugin)
+
+A software product that is managed in the Backstage [Software Catalog](#software-catalog). A component can be a service, website, library, data pipeline, or any other piece of software managed as a single project.
+
+https://backstage.io/docs/features/software-catalog/system-model
+
+## Condition (permission plugin)
+
+Conditions are used to return a conditional decision from a policy. They contain information about a given entity and restrictions on what types of users can view that entity.
```
Comment (dwelsch-esi) [inline docs/references/glossary.md:88]:
```suggestion
A declaration that returns a conditional decision from a policy. Suggestions contain information about a given entity and restrictions on what types of users can view that entity.
```

In general, I recommend trying to make the first sentence of a glossary entry a definition of what the entry *is* rather than what it *does*. You can then expand on the definition if necessary. If the definition becomes too complex, it might be that it's actually documentation that belongs elsewhere in the doc.

File: docs/references/glossary.md [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+
+## Contributor
+
+A volunteer who helps to improve an OSS product such as Backstage. This volunteer effort includes coding, testing, technical writing, user support, and other work. A [user role](#user-role).
+
+## Declarative integration
+
+A new paradigm for Backstage frontend plugins, allowing definition in config files instead of hosting complete React pages.
+
+https://backstage.io/docs/frontend-system
```
Comment (dwelsch-esi) [inline docs/references/glossary.md:102]:
```suggestion
```

File: docs/references/glossary.md [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+
+[Rules](#permission-rules) need additional data before they can be used in a decision. Once a [rule](#permission-rule) is bound to relevant information it forms a [condition](#condition). Conditional decisions tell the [permission framework](#permission) to delegate evaluation to the [plugin](#plugin) that owns the corresponding [resource](#permission-resource). Permission requests that result in a conditional decision are allowed if all of the provided conditions evaluate to be true.
+
+## Contributor
+
+A volunteer who helps to improve an OSS product such as Backstage. This volunteer effort includes coding, testing, technical writing, user support, and other work. A [user role](#user-role).
+
+## Declarative integration
+
+A new paradigm for Backstage frontend plugins, allowing definition in config files instead of hosting complete React pages.
```
Comment (dwelsch-esi) [inline docs/references/glossary.md:100]:
```suggestion
A new paradigm for Backstage frontend plugins, allowing definition in config files instead of hosting complete React pages. See [The Frontend System](https://backstage.io/docs/frontend-system).
```

(Suggest using this syntax for all references back to the documentation. Also, it can go at the end of the definition paragraph.)

File: docs/references/glossary.md [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+
+Someone who writes code and develops software.
+
+A [user role](#user-role) defined as someone who uses a Backstage [app](#app). Might or might not actually be a software developer.
+
+## Developer Portal
+
+A centralized system comprising a user interface and database used to facilitate and document all the software projects within an organization. Backstage is both a developer portal and (by virtue of being based on plugins) a platform for creating developer portals.
+
+## Documents (search plugin)
```
Comment (dwelsch-esi) [inline docs/references/glossary.md:124]:
```suggestion
## Document (search plugin)
```

Comment (dwelsch-esi) [reply docs/references/glossary.md:124]: By convention, glossary entries are singular. But, need to change references too if you do this. Might not be worth it.

File: docs/references/glossary.md [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+
+## CNCF
+
+Cloud Native Computing Foundation.
+
+## Code Grant
+
+[OAuth](#oauth) flow where the client receives an [authorization code](#code) that is passed to the backend to be exchanged for an [access token](#access-token) and possibly a [refresh token](#refresh-token).
+
+## Collators (search plugin)
```
[the remark this answers is not in the document]
Comment (dwelsch-esi) [reply docs/references/glossary.md:76]: Clarification: I see that @freben points out that Wikipedia uses parenthetical qualifiers as well as opening phrases. True. My point here is that in the glossary, one or the other will suffice.

**Review by dwelsch-esi — COMMENTED** [review]
Excellent job on the glossary! I've left some comments on how I would format references, multiple entries, and a couple other things. These are totally optional, though. What you've done is exactly in the spirit of what I recommended. 

I will offer one caution: as you add entries to the glossary, resist the temptation to insert long explanations. The glossary's only function is to define terms -- everything else belongs in the documentation proper.

File: docs/references/glossary.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (sennyeya) [reply docs/references/glossary.md:76]: 👍 , I'm gonna stick with it just in the header and get rid of existing `Part of...` blocks.

File: docs/references/glossary.md (hunk 6 above)
Comment (sennyeya) [reply docs/references/glossary.md:100]: 👍 , I'll have to write a doc about writing for the glossary at this rate :)

Comment (sennyeya) [reply docs/references/glossary.md]: Updated in https://github.com/backstage/backstage/pull/22542/commits/4944f29f2cc61dbd29019cd5693c051301f8497f and added a guide too

File: docs/references/glossary.md (hunk 7 above)
Comment (sennyeya) [reply docs/references/glossary.md]: It's worth the time in the first round to get this right, updated in https://github.com/backstage/backstage/pull/22542/commits/4944f29f2cc61dbd29019cd5693c051301f8497f

File: docs/references/glossary.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (sennyeya) [reply docs/references/glossary.md]: I added some high level rules here for when to add the disambiguator, https://github.com/backstage/backstage/pull/22542/files#diff-f61664ab7d3ef1169d79e285d69cb0231f1f20e989557966a40436946111b9d5R25. Basically, I'm not sure we want them everywhere, but they should be used when the context is not global and there may be confusion between contexts.

File: docs/references/glossary.md (hunk 4 above)
Comment (sennyeya) [reply docs/references/glossary.md]: Definitely agree, this will definitely be something we'll want to iterate on to get the right wording. I updated a couple of the entries that were definitely too wordy. I think pointing back to a plugin or framework doc would help a lot here.

File: docs/references/glossary.md (hunk 6 above)
Comment (dwelsch-esi) [reply docs/references/glossary.md]: Nice!

File: docs/references/glossary.md [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+
+A transform stream that allows you to add additional information to [documents](#document-search-plugin).
+
+## Deployment Artifacts
+
+An executable or package file with all of the necessary information required to deploy the application at runtime. Deployment artifacts can be hosted on [package registries](#package-registry).
+
+## Developer
+
+Someone who writes code and develops software.
```
Comment (dwelsch-esi) [inline docs/references/glossary.md:102]:
```suggestion
1. Someone who writes code and develops software.
```

File: docs/references/glossary.md [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+
+A [user role](#user-role) defined as someone who uses a Backstage [app](#app). Might or might not actually be a software developer.
```
Comment (dwelsch-esi) [inline docs/references/glossary.md:104]:
```suggestion
2. A [user role](#user-role) defined as someone who uses a Backstage [app](#app). Might or might not actually be a software developer.
```

File: docs/references/glossary.md [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+
+https://oauth.net/2/refresh-tokens/
+
+## Resource (catalog plugin)
+
+An [entity](#entity) that represents a piece of physical or virtual infrastructure, for example a database, required by a component. See [the catalog docs](https://backstage.io/docs/features/software-catalog/system-model) for more information.
+
+## Resource (permission plugin)
+
+Not to be confused with [Software Catalog resources](#resource-catalog-plugin). Permission resources represent the objects that users interact with and that can be permissioned.
```
Comment (dwelsch-esi) [inline docs/references/glossary.md:254]:
```suggestion
A representation of an object that a user interacts with and that can be permissioned. Not to be confused with [Software Catalog resources](#resource-catalog-plugin). 
```

File: docs/references/glossary.md [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+
+## Rule (permission plugin)
+
+Rules are predicate-based controls that tap into a [resource](#resource-permission-plugin)'s data.
```
Comment (dwelsch-esi) [inline docs/references/glossary.md:258]:
```suggestion
A predicate-based control that taps into a [resource](#resource-permission-plugin)'s data.
```

File: docs/references/glossary.md [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+
+## Search Engine (Backstage search)
+
+Existing search technology that [Backstage Search](#search) can take advantage of through its modular design. Lunr is the default search in Backstage Search.
+
+## Software Catalog
+
+A Backstage plugin that provides a framework to keep track of ownership and metadata for any number and type of software [components](#component). A core feature of Backstage.
+
+## Software Templates
```
Comment (dwelsch-esi) [inline docs/references/glossary.md:284]: You might want to make this two separate entries. "Software Templates" (plural) is the name of the plugin/feature, so making it singular would be incorrect. "Software Template" is a thing (a project skeleton), and should be listed as a singular noun.

Comment (sennyeya) [reply docs/references/glossary.md:284]: Updated in b408b6e28dbee90afe6a0b76d078dff0f3b39747

**Review by freben — APPROVED** [review]
I say, let's merge and keep working on it!

**Comment by dwelsch-esi** [comment]
@freben @sennyeya @nate-double-u I notice the [glossary issue](https://github.com/backstage/backstage/issues/22001) is still open. Work on the glossary will be ongoing, but the PR is merged and at this point shouldn't we close the issue?
